raiden_contracts/tests/test_channel_unlock.py

```diff
@@ -70,136 +261,272 @@ def test_merkle_root(
+        create_settled_channel,
         reveal_secrets,
 ):
     (A, B) = get_accounts(2)
     settle_timeout = 8
 
+    # Regular channel life-cycle: open -> settle -> unlock -> open -> settle -> unlock
+
     # Mock pending transfers data
-    pending_transfers_tree_A = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
-    reveal_secrets(A, pending_transfers_tree_A.unlockable)
-    unlocked_amount = get_unlocked_amount(
-        secret_registry_contract,
-        pending_transfers_tree_A.packed_transfers,
-    )
+    pending_transfers_tree_1 = get_pending_transfers_tree(web3, [1, 3, 5], [2, 4], settle_timeout)
+    reveal_secrets(A, pending_transfers_tree_1.unlockable)
 
-    # We settle the channel with a smaller locked amount than we will need for the
-    # actual merkle tree of pending transfers
+    # Settle the channel
     create_settled_channel(
         A,
-        pending_transfers_tree_A.locked_amount - 1,
-        pending_transfers_tree_A.merkle_root,
+        pending_transfers_tree_1.locked_amount,
+        pending_transfers_tree_1.merkle_root,
         B,
         0,
         EMPTY_MERKLE_ROOT,
         settle_timeout,
     )
-
-    pre_balance_A = custom_token.functions.balanceOf(A).call()
-    pre_balance_B = custom_token.functions.balanceOf(B).call()
-    balance_contract = custom_token.functions.balanceOf(token_network.address).call()
-    assert balance_contract == pending_transfers_tree_A.locked_amount - 1
-
-    # This should pass, even though the locked amount in storage is smaller.
-    # B will receive less tokens.
     token_network.functions.unlock(
         A,
         B,
-        pending_transfers_tree_A.packed_transfers,
-    ).transact()
+        pending_transfers_tree_1.packed_transfers,
+    ).transact({'from': A})
+
+    # Calling unlock twice does not work
+    with pytest.raises(TransactionFailed):
+        token_network.functions.unlock(
+            A,
+            B,
+            pending_transfers_tree_1.packed_transfers,
+        ).transact({'from': A})
+
+
+def test_unlock_wrong_locksroot(
+        web3,
+        token_network,
+        create_settled_channel,
+        get_accounts,
+        reveal_secrets,
+):
+    (A, B) = get_accounts(2)
+    settle_timeout = 8
+
+    pending_transfers_tree_A = get_pending_transfers_tree(web3, [1, 3, 5], [], settle_timeout)
+    pending_transfers_tree_A_fake = get_pending_transfers_tree(web3, [1, 3, 6], [], settle_timeout)
+    reveal_secrets(A, pending_transfers_tree_A.unlockable)
+    create_settled_channel(
+        A,
+        pending_transfers_tree_A.locked_amount,
+        pending_transfers_tree_A.merkle_root,
+        B,
+        0,
+        EMPTY_MERKLE_ROOT,
+        settle_timeout,
+    )
+
+    with pytest.raises(TransactionFailed):
+        token_network.functions.unlock(
+            A,
+            B,
+            pending_transfers_tree_A_fake.packed_transfers,
+        ).call()
+
+
+def test_channel_unlock_bigger_locked_amount(
```

Comment:
Please reorder the tests as they were, so I can actually see the new changes. I will continue the review afterwards.